bench/main.rs

```diff
@@ -0,0 +1,296 @@
+// Copyright 2019-2021 Tauri Programme within The Commons Conservancy
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-License-Identifier: MIT
+
+use anyhow::Result;
+use serde::Serialize;
+use serde_json::Value;
+use std::{
+  collections::{HashMap, HashSet},
+  env, fs,
+  path::Path,
+  process::{Command, Stdio},
+};
+
+mod utils;
+
+fn read_json(filename: &str) -> Result<Value> {
+  let f = fs::File::open(filename)?;
+  Ok(serde_json::from_reader(f)?)
+}
+
+fn write_json(filename: &str, value: &Value) -> Result<()> {
+  let f = fs::File::create(filename)?;
+  serde_json::to_writer(f, value)?;
+  Ok(())
+}
+
+/// The list of the examples of the benchmark name, arguments and return code
+const EXEC_TIME_BENCHMARKS: &[(&str, &str, Option<i32>)] = &[
+  ("wry_hello_world", "target/release/bench_hello_world", None),
+  (
+    "wry_custom_protocol",
+    "target/release/bench_custom_protocol",
+    None,
+  ),
+  (
+    "wry_cpu_intensive",
+    "target/release/bench_cpu_intensive",
+    None,
+  ),
+];
+
+fn run_strace_benchmarks(new_data: &mut BenchResult) -> Result<()> {
+  use std::io::Read;
+
+  let mut thread_count = HashMap::<String, u64>::new();
+  let mut syscall_count = HashMap::<String, u64>::new();
+
+  for (name, example_exe, _) in EXEC_TIME_BENCHMARKS {
+    let mut file = tempfile::NamedTempFile::new()?;
+
+    Command::new("strace")
+      .args(&[
+        "-c",
+        "-f",
+        "-o",
+        file.path().to_str().unwrap(),
+        utils::root_path().join(example_exe).to_str().unwrap(),
+      ])
+      .stdout(Stdio::inherit())
+      .spawn()?
+      .wait()?;
+
+    let mut output = String::new();
+    file.as_file_mut().read_to_string(&mut output)?;
+
+    let strace_result = utils::parse_strace_output(&output);
+    let clone = strace_result.get("clone").map(|d| d.calls).unwrap_or(0) + 1;
+    let total = strace_result.get("total").unwrap().calls;
+    thread_count.insert(name.to_string(), clone);
+    syscall_count.insert(name.to_string(), total);
+  }
+
+  new_data.thread_count = thread_count;
+  new_data.syscall_count = syscall_count;
+
+  Ok(())
+}
+
+fn run_max_mem_benchmark() -> Result<HashMap<String, u64>> {
+  let mut results = HashMap::<String, u64>::new();
+
+  for (name, example_exe, return_code) in EXEC_TIME_BENCHMARKS {
```

Comment:
Is `return_code` still needed? I see all of them are set to None

Reply:
Yes let's do this. Or maybe comment out the part it supposes to do to remind us.